### ISSUE TITLE
Lazy loading buggy until opening summery window

### ISSUE THREAD

**Issue by weilbith** [rcarriga/vim-ultest#27]
**Describe the bug**
I'm a big fan of loading plugins lazy when they become necessary (e.g. I don't need this plugin until I actually want to run a test or for writing an email not at all).
But apparently this does not work. The plugin gets loaded and the commands etc are their. But if I run `:Ultest` or `:UltestNearest`, nothing happens. Silence. I added some debugging messages to the VimL stuff. And I can trace it down to [this call](https://github.com/rcarriga/vim-ultest/blob/40165b1201fefd2e7ae13f1884c120dec850ed4c/autoload/ultest/handler.vim#L13) where it switches to Python. In python I don't anything to print. No clue what is happening here. If I execute `:echo call('_ultest_run_all', ['path/to/file'])` it always just prints `1`. So the plugin is loaded and it basically does something. Just that nothing happens...
When I open the summary window with `:UltestSummary` and then go to a test case and hit `r` to execute the case, it works. From now on I can also call `:Ultest` and `:UltestNearest`. 
If I turn of the lazy loading of the plugin and always source it at startup, it just works and I can call `:Ultest` right away.

Logs should be created using `ULTEST_LOG_LEVEL=DEBUG ULTEST_LOG_FILE=vim-ultest.log nvim <test-file>` and added here as well

```
11:36:06 | INFO | MainThread | logging.py:create_logger:71 | Logger created
11:36:06 | DEBUG | MainThread | __init__.py:__init__:40 | Handler created
11:36:06 | DEBUG | Thread-1 | jobs.py:_handle_coroutine:42 | Starting job with group run_all
11:36:06 | INFO | Thread-1 | __init__.py:run:101 | Running all tests in src/main.rs
11:36:06 | DEBUG | Thread-1 | jobs.py:_handle_coroutine:57 | Finished job with group run_all
```

**To Reproduce**

This assumes that `vim-test` is always available in path (simplification for the moment of testing). So this is the `packer.nvim` configuration:

```vim
lua packer.use({
      \   'rcarriga/vim-ultest',
      \   run = ':UpdateRemotePlugins',
      \   cmd = {'Ultest', 'UltestNearest', 'UltestSummary'},
      \ })
```

**Expected behavior**

The plugin works after it got loaded and you can execute `:Ultest`.

**Comment by rcarriga** [comment]
I've added in support for lazy loading though you will still find that the first time you run a command it will not work. This is because the processing of the test positions happens asynchronously and so has not happened when you run the command initially.

The second time you run a command it will work though.

**Comment by weilbith** [comment]
Hmm. This still does not work for me. Is there maybe a function I can call manually on plugin souring that will do the processing you have mentioned?

**Comment by rcarriga** [comment]
Hmm that's odd, can you post new logs. Also what NeoVim version are you running?

**Comment by rcarriga** [comment]
I'd rather have the functionality be automatic than relying on users needing to run extra functions. It works perfectly for me with the packer config you posted so it is likely something small not working for you

**Comment by weilbith** [comment]
Okay here we go.
My NeoVim version is from the AUR nightly bin package (so very recent):
```
NVIM v0.5.0-dev+1160-g0f7b6b2ef
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/runner/work/neovim/neovim/build/config -I/home/runner/work/neovim/neovim/src -I/home/runner/work/neovim/neovim/.deps/usr/include -I/usr/include -I/home/runner/work/neovim/neovim/build/src/nvim/auto -I/home/runner/work/neovim/neovim/build/include
Compiled by runner@fv-az166-971

Features: +acl +iconv +tui
```

My current checked out version of `vim-ultest` is 10ffa74732a7e422afbc1415a1c0d510af2d8c5b
I'm currently working on wbthomason/packer.nvim#264 to my `packer.nvim` is on `bc121b95126dce9f83e5d275c46569da2542d028`. This bascially just resolves the issue that when ever `vim-ultest` is loaded (lazy), `vim-test` gets loaded before it has not been loaded yet.

The now exact `packer.nvim` configuration is the following (note that I have brought `packer` into the global Lua scope [required for my extensive package usage]):

```vim
lua packer.use({
      \   'janko/vim-test',
      \   cmd = { 'TestNearest', 'TestFile', 'TestSuite', 'TestLast', 'TestVisit' },
      \   config = 'vim.cmd("packadd test")',
      \ })
    
" and somewhere else...

lua packer.use({
      \   'rcarriga/vim-ultest',
      \   run = ':UpdateRemotePlugins',
      \   wants = {'vim-test'},
      \   cmd = {'Ultest', 'UltestNearest', 'UltestSummary'},
      \   config = 'vim.cmd("packadd ultest")',
      \ })
```

The configuration for the plugins is inside the packages which get loaded via the `config` properties (tell me if you need insights here).

Now here is a simple Rust file with a test:

```rs
pub fn divide(dividend: u32, divisor: u32) -> u32 {
    if divisor == 0 {
        panic!("Division by zero!");
    }

    dividend / divisor
}

#[cfg(test)]

mod tests {
    use super::{divide};
    #[test]
    fn test_divide() {
        assert_eq!(divide(4, 2), 2);
    }
    
    #[test]
    #[should_panic]
    fn test_divide_zero_divisor() {
        divide(3, 0);
    }
}
```

And thats it.

**Comment by rcarriga** [comment]
OK I was able to reproduce. The reason why it's inconsistent is that there is a bug in pynvim which causes issues with lazy loading. I seem to have fixed it but I will wait for you to confirm with latest commit

**Comment by weilbith** [comment]
Thank you very much for taking the time! 🙏🏾  The lazy loading part seem to work fine. 🤗 
The only remaining issue is that the first command always fails. You said this is related to some initial asynchronous parsing. Is this parsing done in a function? If so I could easily call this function myself up on loading before the command executes. Personally I would have no issue with that.

**Comment by rcarriga** [comment]
Brilliant!

It's not that the async function needs to be called, that's already done once the plugin is loaded. The problem is that the async function needs to complete before you can execute a command because the command relies on the test positions already being parsed. 

I could add in the functionality for forcing the the test  detection when the command runs but it would require a more substantial change. I'll do that when I have the time. For now there is a user autocmd `UltestPositionsUpdate` that is triggered when the file is processed. You could use this somehow to re-run your command with something like:
```vim
  au User UltestPositionsUpdate ++once UltestNearest
```
though obviously you'd need to be able to tell which command was run

**Comment by weilbith** [comment]
Okay. Thanks for explaining me the details! 🙃

**Comment by rcarriga** [comment]
I'll leave this open and close it once I've implemented the forced update. I don't think it's fair to say that lazy loading works well until then

**Comment by rcarriga** [comment]
OK now should be working without any issue, just like normal loading.
